instrumentation/reactor-3.1/src/test/groovy/ReactorCoreTest.groovy

```diff
@@ -242,117 +241,59 @@ class ReactorCoreTest extends AgentTestRunner {
     when:
-    runUnderTrace {
-      // The "add one" operations in the publisher created here should be children of the publisher-parent
-      Publisher<Integer> publisher = publisherSupplier()
-
-      def tracer = OpenTelemetry.getTracer("test")
-      def intermediate = tracer.spanBuilder("intermediate").startSpan()
-      // After this activation, the "add two" operations below should be children of this span
-      def scope = tracer.withSpan(intermediate)
-      try {
-        if (publisher instanceof Mono) {
-          return ((Mono) publisher).map(addTwo)
-        } else if (publisher instanceof Flux) {
-          return ((Flux) publisher).map(addTwo)
-        }
-        throw new IllegalStateException("Unknown publisher type")
-      } finally {
-        intermediate.end()
-        scope.close()
-      }
+    def mono = Mono.just(42)
+      .map(addOne)
+      .map(addTwo)
+    TraceUtils.runUnderTrace("trace-parent") {
+      mono.block()
     }
 
     then:
     assertTraces(1) {
-      trace(0, (workItems * 2) + 3) {
-        span(0) {
-          operationName "trace-parent"
-          parent()
-          attributes {
-          }
-        }
-
-        basicSpan(it, 1, "publisher-parent", span(0))
-        basicSpan(it, 2, "intermediate", span(1))
-
-        for (int i = 0; i < workItems; i++) {
-          span(3 + i) {
-            operationName "add two"
-            childOf span(2)
-            attributes {
-            }
-          }
-        }
-        for (int i = 0; i < workItems; i++) {
-          span(3 + workItems + i) {
-            operationName "add one"
-            childOf span(1)
-            attributes {
-            }
-          }
-        }
+      trace(0, 3) {
+        basicSpan(it, 0, "trace-parent")
+        basicSpan(it, 1, "add one", span(0))
+        basicSpan(it, 2, "add two", span(0))
       }
     }
 
-    where:
-    name         | workItems | publisherSupplier
-    "basic mono" | 1         | { -> Mono.just(1).map(addOne) }
-    "basic flux" | 2         | { -> Flux.fromIterable([1, 2]).map(addOne) }
   }
 
-  def "Publisher chain spans can have the parent removed at assembly time '#name'"() {
+  def "Publisher chain spans have the correct parents from subscription time '#name'"() {
     when:
     runUnderTrace {
-      // The operations in the publisher created here all end up children of the publisher-parent
+      // The "add one" operations in the publisher created here should be children of the publisher-parent
       Publisher<Integer> publisher = publisherSupplier()
 
-      // After this activation, all additions to the assembly will create new traces
       def tracer = OpenTelemetry.getTracer("test")
-      def scope = tracer.withSpan(DefaultSpan.getInvalid())
+      def intermediate = tracer.spanBuilder("intermediate").startSpan()
+      // After this activation, the "add two" operations below should be children of this span
```

Comment:
This comment doesn't match the assertions below. Not sure that is intended or not